### PR TITLE
Add prometheus metrics endpoint

### PR DESCRIPTION
**Pull request by 0xSte**
## 📝 Summary

An initial implementation of some prometheus metrics for MEV Boost

## ⛱ Motivation and Context

I want to be able to monitor my mev-boost container with prometheus to get better operational insight

## 📚 References

https://github.com/flashbots/mev-boost/issues/253
---

## ✅ I have run these commands

* [✔️] `go mod tidy`
* [✔️] `make test`
* [✔️] `make lint `
* [❔] `make run-mergemock-integration`

Getting an error back from `make run-mergemock-integration`, but not sure if it's related to my change
```
ERROR  [2022-09-05T19:09:20+01:00] Unable to retrieve proposal payload           error="builder REST API returned non-200 status code: 204" slot="5"
make: *** [run-mergemock-integration] Error 1```

### CODE REVIEW

File: server/relay_entry_test.go [hunk 1]
```diff
@@ -45,23 +45,23 @@ func TestParseRelaysURLs(t *testing.T) {
 			expectedURL:       "http://" + publicKey.String() + "@foo.com",
 		},
 		{
-			name:     "Relay URL with public key host and port",
+			name:     "Relay URL with public key host and Port",
```
Comment (metachris) [inline server/relay_entry_test.go:48]: why?

Comment (stefanomantini) [reply server/relay_entry_test.go:48]: really not sure... maybe something weird as part of a save actions rule i've got setup in in goland i think... will get this revised 👍

File: server/metrics.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	"github.com/prometheus/client_golang/prometheus/promauto"
+)
+
+const (
+	metricNamespace = "mev"
+	metricSubsystem = "boost"
+)
+
+var (
+	metricHTTPRequestsSent = promauto.NewCounterVec(prometheus.CounterOpts{
```
[the remark this answers is not in the document]
Comment (stefanomantini) [reply server/metrics.go]: was supposed to refactor this before pushing the original PR, should be sorted in latest

File: server/service.go [hunk 3]
```diff
@@ -162,13 +203,22 @@ func (m *BoostService) StartHTTPServer() error {
...
+			}
+			labels := map[string]string{
+				"slot":   fmt.Sprintf("%d", k.slot),
+				"pubkey": bidResp.response.Data.Message.Pubkey.PubkeyHex().String(),
 			}
+			m.metricsMev.gasUsed.With(labels).Set(float64(bidResp.response.Data.Message.Header.GasUsed))
+			m.metricsMev.gasLimit.With(labels).Set(float64(bidResp.response.Data.Message.Header.GasLimit))
+			m.metricsMev.timeSinceBid.With(labels).Set(float64(time.Since(bidResp.t).Milliseconds()))
 		}
+		m.metricsMev.bids.Set(float64(len(m.bids)))
```
Comment (stefanomantini) [inline server/service.go:221]: @metachris i'm not particularly intimate with the bidding process, but had a rough stab at how i think we could export metrics for this. This seemed like the most central place we'd be able to emit metrics from (albeit 1m delayed)

Would be great to hear your thoughts on this

Comment (stefanomantini) [reply server/service.go:221]: also not too sure why the `TestGetPayload/Bad_response_from_relays` test is failing, will look into it after coffee

Comment (stefanomantini) [reply server/service.go:221]: tests fixed

**Comment by ybstaked** [comment]
what's the progress on this?  we also would appreciate having prometheus metrics on boost

File: README.md [hunk 4]
```diff
@@ -212,6 +212,38 @@ Run mev-boost pointed at our [Sepolia Relay](https://builder-relay-sepolia.flash
...
+# Metrics
+
+Optionally the `-prometheus=true` flag can be passed to the binary to collect various prometheus metrics
+
+## MEV Metrics
+Metrics relating to mev specifics
+
+| **metric**                     | **type** | **labels** | **description**                             |
+|--------------------------------|----------|------------|---------------------------------------------|
+| mev_boost_validator_info       | gauge    | pubkey     | validator identity information              |
```
Comment (ybstaked) [inline README.md:224]: wouldn't having a number of validators successfully registered per relay be more helpful?  per pubkey might be a lot of data for validators operating at scale

**Comment by metachris** [comment]
These changes are deep and extensive into the core system and logic. Considering that mev-boost is run at 50% of all validators, it is too risky and impactful for adding a large amount of new logic in the core flows at the current time.

Perhaps consider collecting the metrics from the logs, or thinking about a possible plugin interface. 🤔

**Comment by 0xSte** [comment]
> These changes are deep and extensive into the core system and logic. Considering that mev-boost is run at 50% of all validators, it is too risky and impactful for adding a large amount of new logic in the core flows at the current time.
> 
> Perhaps consider collecting the metrics from the logs, or thinking about a possible plugin interface. 🤔

Thanks for your feedback @metachris, i may pick this up again later and try coming at it from a different angle 👍

**Comment by ybstaked** [comment]
> 



> These changes are deep and extensive into the core system and logic. Considering that mev-boost is run at 50% of all validators, it is too risky and impactful for adding a large amount of new logic in the core flows at the current time.
> 
> Perhaps consider collecting the metrics from the logs, or thinking about a possible plugin interface. 🤔

@metachris as this PR is implemented as is, seems this would've been behind a `-metrics` flag (similar to prysm and teku have their prometheus metrics behind a flag on a separate metrics port).  Making the metrics opt-in and improving the metrics collected, I think this would be a useful feature to add native prometheus metrics

**Comment by 0xSte** [comment]
@ybstaked @metachris happy to reopen and rebase this PR if there's any will to get it implemented

Otherwise the suggestion is the approach outlined in #370

**Comment by metachris** [comment]
I sincerely appreciate the effort in adding the prometheus support. As mentioned before, the changes reach too deep into the core program flow to merge and force on a majority of proposers, without extensive review and in particular discussions beforehand. If you feel you'd like to pursue this, then please open an issue and continue the conversation there.

**Pull request closed without merging**